src/NHibernate/Collection/Generic/SetHelpers/SetSnapShot.cs

```diff
@@ -1,109 +1,178 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
+#if NETCOREAPP2_0
+using System.Runtime.Serialization;
+using System.Threading;
+#endif
 
 namespace NHibernate.Collection.Generic.SetHelpers
 {
+#if NETFX || NETSTANDARD2_0
+	// TODO 6.0: Consider removing this class in case we upgrade to .NET 4.7.2 and NET Standard 2.1,
+	// which have HashSet<T>.TryGetValue
 	[Serializable]
-	internal class SetSnapShot<T> : ISetSnapshot<T>
+	internal class SetSnapShot<T> : ICollection<T>, IReadOnlyCollection<T>, ICollection
```

Comment:
Deserialization callback is missing which is required to properly deserialize Dictionary. I think this was a part of a reason why it was implemented using `List<>` in a first place.

Reply:
Maybe I am missing something as I don't understand why we need a deserialization callback to properly deserialize the dictionary and what should the callback do. I've [made a fiddle](https://dotnetfiddle.net/evzfmL) to test the serialization of `SetSnapShot` and it's working without the callback. Also `PersistentGenericMap` has a dictionary field and it doesn't implement `IDeserializationCallback`, so if we need the callback for `SetSnapShot`, then we probably would need to add it for `PersistentGenericMap`.

Reply:
Is it about #1891? I believe it is only the case if hashcodes are cached for dictionary keys

Reply:
As far as I remember, the issue with dictionaries happens when the type may be deserialized from a graph of entities which own deserialization requires having the deserialized dictionaries populated early.
The concrete example of this need is a cache key class which "value" depends on an internal dictionary, while the key has been serialized as part of another dictionary. In such case we must ensure the cache key own internal dictionary is deserialized and ready for key computing before the key is added to its containing dictionary.
Implementing `IDeserializationCallback` is then required because we must then serialize the dictionary as a simpler type (an array), available sooner in the deserialization process.

For the case here, I do not think we need to do this. This `SetSnapShot` class does not need to have its state ready early for being correctly deserialized inside a graph of entity, as its state is not required during the deserialization process of the entities it may belong to.